staging/src/k8s.io/dynamic-resource-allocation/resourceslice/tracker/tracker.go

```diff
@@ -0,0 +1,539 @@
+	// the event handlers are allowed to access the cache. Holding rwMutex
+	// then would cause a deadlock.
+	//
+	// New functions get added as part of processing a cache update while
+	// the rwMutex is locked. Each function which adds something to the queue
+	// also drains the queue before returning, therefore it is guaranteed
+	// that all event handlers get notified immediately (useful for unit
+	// testing).
+	//
+	// A channel cannot be used here because it cannot have an unbounded
+	// capacity. This could lead to a deadlock (writer holds rwMutex,
+	// gets blocked because capacity is exhausted, reader is in a handler
+	// which tries to lock the rwMutex). Writing into such a channel
+	// while not holding the rwMutex doesn't work because in-order delivery
+	// of events would no longer be guaranteed.
+	eventQueue queue.FIFO[func()]
+}
+
+func StartTracker(ctx context.Context, kubeClient kubernetes.Interface, informerFactory informers.SharedInformerFactory) (*Tracker, error) {
+	t := newTracker(informerFactory)
+	return t, t.start(ctx, kubeClient)
+}
+
+func newTracker(informerFactory informers.SharedInformerFactory) *Tracker {
+	t := &Tracker{
+		resourceSlices:        informerFactory.Resource().V1beta1().ResourceSlices().Informer(),
+		resourceSlicePatches:  informerFactory.Resource().V1alpha3().ResourceSlicePatches().Informer(),
+		deviceClasses:         informerFactory.Resource().V1beta1().DeviceClasses().Informer(),
+		celCache:              cel.NewCache(10), // TODO: share cache with scheduler
+		patchedResourceSlices: cache.NewStore(cache.MetaNamespaceKeyFunc),
+	}
+	t.handlerRegistration = handlerRegistrationFunc(func() bool {
+		return t.resourceSlices.HasSynced() &&
+			t.resourceSlicePatches.HasSynced() &&
+			t.deviceClasses.HasSynced()
+	})
+	return t
+}
+
+func (t *Tracker) start(ctx context.Context, kubeClient kubernetes.Interface) error {
+	err := t.initInformers(ctx)
+	if err != nil {
+		return err
+	}
+	// kubeClient is not always set in unit tests.
+	if kubeClient != nil {
+		eventBroadcaster := record.NewBroadcaster(record.WithContext(ctx))
+		eventBroadcaster.StartLogging(klog.Infof)
+		eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: kubeClient.CoreV1().Events("")})
+		t.recorder = eventBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: "resource_slice_tracker"})
+	}
+	return nil
+}
+
+func (t *Tracker) ListPatchedResourceSlices() ([]*resourceapi.ResourceSlice, error) {
+	return typedSlice[*resourceapi.ResourceSlice](t.patchedResourceSlices.List()), nil
+}
+
+// AddEventHandler adds an event handler to the tracker. Events to a
+// single handler are delivered sequentially, but there is no
+// coordination between different handlers. A handler may use the
+// tracker.
+//
+// The return value can be used to wait for cache synchronization.
+func (t *Tracker) AddEventHandler(handler cache.ResourceEventHandler) cache.ResourceEventHandlerRegistration {
```

Comment:
All of the `AddEventHandler`-adjacent things here are mostly copy-paste from the scheduler's `AssumeCache`: https://github.com/kubernetes/kubernetes/blob/be80fec686dbe288b0c254a3413e5fe16150fe44/pkg/scheduler/util/assumecache/assume_cache.go#L488